src/Common/JemallocNodumpAllocatorImpl.cpp

```diff
@@ -0,0 +1,84 @@
+#include <Common/JemallocNodumpAllocatorImpl.h>
+
+#if USE_JEMALLOC
+#include <Poco/Format.h>
+#include <base/errnoToString.h>
+#include <base/defines.h>
+#include <sys/mman.h>
+#include <cstring>
+
+namespace DB
+{
+
+JemallocNodumpAllocatorImpl & JemallocNodumpAllocatorImpl::instance()
+{
+    static JemallocNodumpAllocatorImpl allocator;
+    return allocator;
+}
+
+void * JemallocNodumpAllocatorImpl::allocate(size_t size, size_t alignment) const
+{
+    return mallocx(size, alignment ? flags | MALLOCX_ALIGN(alignment) : flags);
+}
+
+void JemallocNodumpAllocatorImpl::deallocate(void * p) const
+{
+    dallocx(p, flags);
+}
+
+void * JemallocNodumpAllocatorImpl::alloc(
+    extent_hooks_t * extent, void * new_addr, size_t size, size_t alignment, bool * zero, bool * commit, unsigned arena_ind)
+{
+    void * result = original_alloc_(extent, new_addr, size, alignment, zero, commit, arena_ind);
+    if (result != nullptr)
+    {
+#if defined(MADV_DONTDUMP)
+        if (auto ret = madvise(result, size, MADV_DONTDUMP))
+        {
+            throw std::runtime_error(Poco::format("Failed to run madvise: %s", errnoToString(ret)));
+        }
+#elif defined(MADV_NOCORE)
+        if (auto ret = madvise(result, size, MADV_NOCORE))
+        {
+            throw std::runtime_error(Poco::format("Failed to run madvise: %s", errnoToString(ret)));
+        }
+#endif
+    }
+    return result;
+}
+
+JemallocNodumpAllocatorImpl::JemallocNodumpAllocatorImpl()
+{
+    setupArena();
+}
+
+void JemallocNodumpAllocatorImpl::setupArena()
+{
+    size_t len = sizeof(arena_index);
+    if (auto ret = mallctl("arenas.create", &arena_index, &len, nullptr, 0))
+    {
+        throw std::runtime_error(Poco::format("Failed to create jemalloc arena: %s", errnoToString(ret)));
```

Comment:
Is there any strong reason to use `Poco::Format` instead of `fmt::format`?

Reply:
Moved to `fmt::format`